File: abc298/abc298_c.py
from collections import defaultdict


def main():
    N = int(input())
    Q = int(input())

    card = defaultdict(set)
    box = defaultdict(list)

    for _ in range(Q):
        tmp = list(map(int, input().split(" ")))
        query = tmp[0]
        if query == 1:
            i = tmp[1]
            j = tmp[2]
            card[i].add(j)
            box[j].append(i)
        elif query == 2:
            i = tmp[1]
            ans = box[i]
            ans.sort()
            print(*ans)
        else:
            i = tmp[1]
            ans = list(card[i])
            ans.sort()
            print(*ans)


main()

File: abc298/abc298_c.cpp
#include <bits/stdc++.h>
using namespace std;

int main()
{
	int N, Q;
	cin >> N >> Q;

	map<int, multiset<int>> box;
	map<int, set<int>> card;

	while (Q--)
	{
		int query;
		cin >> query;

		if (query == 1)
		{
			int i, j;
			cin >> i >> j;

			card[i].insert(j);
			box[j].insert(i);
		}
		else if (query == 2)
		{
			int i;
			cin >> i;

			for (auto x : box[i])
			{
				cout << x << " ";
			}
			cout << endl;
		}
		else
		{

			int i;
			cin >> i;

			for (auto x : card[i])
			{
				cout << x << " ";
			}
			cout << endl;
		}
	}

	return 0;
}
